Refinements (#3)
degen-lab/stacks-mocking-functions: README.md

# wrapper-testing-bitcoin-and-stacks [![npm](https://img.shields.io/npm/v/wrapper-testing-bitcoin-and-stacks?color=red)](https://www.npmjs.com/package/wrapper-testing-bitcoin-and-stacks)

> This is a TypeScript library designed to simplify the process of mocking various functionalities for testing and development purposes. The library offers a suite of helper functions to mock different scenarios.

Reviewer: ```suggestion
This library provides a convenient wrapper for testing and developing on Bitcoin and Stacks. It offers helper functions to mock different scenarios.
```

  - [validateAddress](#validateaddress)
  - [wifToPublicKey](#wiftopublickey)
  - [wifToPrivateKey](#wiftoprivatekey)
  - [mnemonicToAddress](#mnemonictoaddress)
  - [mnemonicToPubKey](#mnemonictopubkey)
  - [mnemonicToPrivKey](#mnemonictoprivkey)

## PoxHelpers

### currentRewardCycleMock

This function is used for getting the current reward cycle of the given Simnet instance.

```typescript
import { currentRewardCycleMock } from '@something/mocking';
import { Simnet, initSimnet } from '@hirosystems/clarinet-sdk';

let simnet: Simnet;
simnet = await initSimnet();

const address = 'ST1PQHQKV0RJXZFY1DGX8MNSNYVE3VGZJSRTPGZGM'; // STX address

const currentRewardCycle = currentRewardCycleMock(simnet, address);
```

### burnHeightToRewardCycleMock

This function is used for identifying what cycle a given burn height belongs to for a given Simnet instance.

```typescript
import { burnHeightToRewardCycleMock } from '@something/mocking';
import { Simnet, initSimnet } from '@hirosystems/clarinet-sdk';

let simnet: Simnet;
simnet = await initSimnet();

const address = 'ST1PQHQKV0RJXZFY1DGX8MNSNYVE3VGZJSRTPGZGM'; // STX address
const height = 2050; //any number

> const rewadrCycle = burnHeightToRewardCycleMock(simnet, address, height);

Reviewer: ```suggestion
const rewardCycle = burnHeightToRewardCycleMock(simnet, address, height);
```

```

### rewardCycleToBurnHeightMock

This function is used for getting the burn height that a cycle starts at for a given Simnet instance.

```typescript
import { rewardCycleToBurnHeightMock } from '@something/mocking';
import { Simnet, initSimnet } from '@hirosystems/clarinet-sdk';

let simnet: Simnet;
simnet = await initSimnet();

const address = 'ST1PQHQKV0RJXZFY1DGX8MNSNYVE3VGZJSRTPGZGM'; // STX address
const rewCycle = 2; //any number

const burnHeight = rewardCycleToBurnHeightMock(simnet, address, rewCycle);
```

> ### getPoxInfoMock

Reviewer: do we want to get other static data from the api? like activation height for different pox contracts (pox-2, pox3, pox4)
or dinamic values? like `blocks_until_reward_phase`, `blocks_until_prepare_phase`

```typescript
import { customBlocksUntilPreparePhaseMock } from '@something/mocking';
import { Simnet, initSimnet } from '@hirosystems/clarinet-sdk';

const blockHeight = 2005;

const blocksUntilRew = customBlocksUntilPreparePhaseMock(blockHeight);
```

Also there can be entered custom `rewardCycleLength`, `prepareCycleLength`, `startBurnHeight`:

```typescript
import { customBlocksUntilPreparePhaseMock } from '@something/mocking';
import { Simnet, initSimnet } from '@hirosystems/clarinet-sdk';

const blockHeight = 2005;
const rewardCycleLength = 40;
const prepareCycleLength = 500;
const startBurnHt = 100;

const blocksUntilPrepare = customBlocksUntilPreparePhaseMock(
  blockHeight,
  rewardCycleLength,
  prepareCycleLength,
  startBurnHt
);
```

### isInPreparePhaseMock

This function is used for checking if the given burn height is a prepare phase block of a given Simnet instance, returning a boolean.

```typescript
import { isInPreparePhaseMock } from '@something/mocking';
import { Simnet, initSimnet } from '@hirosystems/clarinet-sdk';

let simnet: Simnet;
simnet = await initSimnet();

const address = 'ST1PQHQKV0RJXZFY1DGX8MNSNYVE3VGZJSRTPGZGM'; // STX address

> const burnHeigth = 2059;

Reviewer: ```suggestion
const burnHeight = 2059;
```

let simnet: Simnet;
simnet = await initSimnet();

const liquidSupply = stxLiquidSupply(simnet);
```

### stxAccount

This function queries the Simnet instance to retrieve the details of the specified STX account as a javascript dictionary:

- 'locked': The locked amount in ustx.
- 'unlock-height': The height where the locked amount will be unlocked.
- 'unlocked': The unlocked amount in ustx.

```typescript
import { stxAccount } from '@something/mocking';
import { Simnet, initSimnet } from '@hirosystems/clarinet-sdk';

let simnet: Simnet;
simnet = await initSimnet();

const address = 'ST1PQHQKV0RJXZFY1DGX8MNSNYVE3VGZJSRTPGZGM'; // STX address

const account = stxAccount(simnet, address);
```

### getDataVar

This function queries the Simnet instance to retrieve the value of a specific data variable from the specified contract.

```typescript
import { getDataVar } from '@something/mocking';
import { Simnet, initSimnet } from '@hirosystems/clarinet-sdk';

let simnet: Simnet;
simnet = await initSimnet();

const contract = 'your-contract';
const varName = 'your-data-variable';

const dataVar = getDataVar(simnet, contract, varName);
```

> ### getMapEntry

Reviewer: can this also be used on networks like Devnet, Testnet, Mainnet or only usable on simnet?

Author: This can only be used on simnet as it is using the getDataVar function of a Simnet instance, it would need a different implementation to use other networks.

This function queries the Simnet instance to retrieve the entry of a specific map, identified by its name and key (as ClarityValue), from the specified contract.

```typescript
import { getMapEntry } from '@something/mocking';
import { Simnet, initSimnet } from '@hirosystems/clarinet-sdk';
import { Cl } from '@stacks/transactions';

let simnet: Simnet;
simnet = await initSimnet();

const contract = 'your-contract';
const mapName = 'your-map-name';

> const mapKey = Cl.tuple({ 'your-key': your - value });

Reviewer: ```suggestion
const mapKey = Cl.tuple({ 'your-key': 'your-value' });
```
